privacy-protections/fingerprinting/canvas.js

```diff
@@ -0,0 +1,194 @@
+    async () => {
+        const limit = 200;
+        const canvasElement = createCanvas(3840, 2160);
+        const canvasContext = canvasElement.getContext('2d');
+        const before = performance.now();
+        const canvasData = canvasContext.getImageData(0, 0, canvasElement.width, canvasElement.height);
+        const after = performance.now();
+        const time = after - before;
+        ok(time < limit, `Getting image data must be under ${limit}ms (${time})`);
+
+        for (let i = 0; i < canvasData.data.length; i++) {
+            canvasData.data[i] = Math.floor(Math.random() * 256);
+        }
+        canvasContext.putImageData(canvasData, 0, 0);
+
+        const before2 = performance.now();
+        const canvasData2 = canvasContext.getImageData(0, 0, canvasElement.width, canvasElement.height);
+        const after2 = performance.now();
+        const time2 = after2 - before2;
+        ok(time2 < limit, `Getting image data must be under ${limit}ms (${time2})`);
+
+        // Validate canvas caching didn't return blank pixels here
+        // We can't check for equality as the browser may modify on put.
+        ok(canvasData2.data.find((v) => v !== 0), 'Data in image must not be blank');
+    },
+
+    // Load a known pixel image and check the values when rendered to the canvas
+    async () => {
+        const limit = 15;
+        const image = document.createElement('img');
+        image.src = './pixel.gif';
+        let res;
+        const promise = new Promise((resolve) => {
+            res = resolve;
+        });
+        image.onload = (e) => {
+            const refImage = e.target;
+            const canvasElement = createCanvas(refImage.width, refImage.height);
+            const canvasContext = canvasElement.getContext('2d');
+            canvasContext.drawImage(refImage, 0, 0);
+            const before = performance.now();
+            const canvasData = canvasContext.getImageData(0, 0, canvasElement.width, canvasElement.height);
+            const after = performance.now();
+            eq(canvasData.data[0], 18, 'Red should be 18');
+            eq(canvasData.data[1], 18, 'Blue should be 18');
+            eq(canvasData.data[2], 255, 'Green should be 255');
+            eq(canvasData.data[3], 255, 'Alpha should be 255');
+            const time = after - before;
+            ok(time < limit, `Getting image data must be under ${limit}ms (${time})`);
+            const knownOutputUrls = [
+                // Firefox
+                'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVQImWMQEvr/HwADgAIj6OBrXgAAAABJRU5ErkJggg==',
+
+                // Chrome
+                'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAAAXNSR0IArs4c6QAAAA1JREFUGFdjEBL6/x8AA4ACI+VyxooAAAAASUVORK5CYII='
+            ];
+            ok(knownOutputUrls.includes(canvasElement.toDataURL()), 'Output conversion has expected values');
```

Comment:
there is no safari value predefined here and it fails on Safari

BTW do we know if those prefefined values are the same on non-mac's?